fiducial_follow/nodes/follow.py

```diff
@@ -58,20 +69,74 @@ def __init__(self):
        self.br = tf2_ros.TransformBroadcaster()
 
-       # A publisher for robot motion commands
-       self.cmdPub = rospy.Publisher("/cmd_vel", Twist, queue_size=1)
 
        # Flag to avoid sending repeated zero speeds
        self.suppressCmd = False
 
-       # The name of the coordinate frame of the fiducial we are interested in
+       # define a move time that when non-zero we continue to do rotate or drive commands
+       self.moveTimeLeft = 0.0
+
+       # ---------------------- start of keywords for use in commands ---------------------------
+
+       # define a fiducial that when set implies we are not seeking any fiducial
+       self.null_fiducial = "fid0"
+
+       # The default fiducial we will start to follow from the start. Use of null_fiducial means idle at start
+       # The legacy fiducial follow used marker 49 so default to that here
+       # Set this to self.null_fiducial if you want idle at start awaiting commands
        self.target_fiducial = rospy.get_param("~target_fiducial", "fid49")
 
+       # optionally we can disable verbose debug messages or leave them on as in legacy code
+       # these were on in legacy version but can be disabled using parameter below
+       self.debug_follow = rospy.get_param("~debug_follow", True)
+
+       # Set to True for Legacy follower where we search for the sole fiducial if it goes out of sight
+       # to operate in the mode where commands are used via topic this should be False
+       self.search_for_target = rospy.get_param("~search_for_target", True)
+
+       # ---------------------- start of keywords for use in commands ---------------------------
+
+       # Defines that are of value to pass in as elements of messages in the inbound command topic
+
+       # default action we will use once we actually find and have approached the fiducial
+       # The awaitNextCommand action will continue to follow until a new command comes in
+       self.actOnDoneStopFollowing = "StopFollowing"
+       self.actOnDoneDoNextCommand = "DoNextCommand"
+       self.actOnDone = self.actOnDoneDoNextCommand
+
+       # We need a state to indicate for the current command so we define 'InProgress' and 'Done'
+       # When cmdStatus is InProgress we do not read new command except state control commands
+       self.cmdStatusInProgress = "InProgress"
+       self.cmdStatusDone       = "Done"
+       self.cmdStatus = self.cmdStatusDone
+
+       # define names for specific commands 
+       self.cmdFollowFiducial  = "FollowFiducial"
+       self.cmdClearCommands   = "ClearCommands"
+       self.cmdClearInProgress = "ClearInProgress"
+       self.cmdStopMovement    = "StopMovement"
+       self.cmdDriveForward    = "DriveForward"
+       self.cmdDriveReverse    = "DriveReverse"
+       self.cmdSetDriveRate    = "SetDriveRate"
+       self.cmdRotateLeft      = "RotateLeft"
+       self.cmdRotateRight     = "RotateRight"
+       self.cmdSetRotateRate   = "SetRotateRate"
```

Comment:
Could we have a list for these, instead of lots of variables?

Reply:
The functionality in need is to have the string appear in only one place that means a specific command.  Here I was doing the equivalent of a define.  What  I would do in C would use an array of structs and in C++ a map container.    In Python I don't know the ideal approach but I could use a  list of tuples to satisfy your request you make where each tuple is for a specific command which is not elegant but would have minimal risk of munging up the already debugged code and would avoid hard coded strings.  There may be a 'container' in python that can be accessed by a string and if no 'hit' you knew it was not in that container.

Reply:
It might be better to have module level variables than class instance variables for this purpose. You could also use an enum or a dictionary instead of individual variables.

Reply:
Python dictionaries are like C++ maps, but more friendly